tools/sst_rewrite.cc

```diff
@@ -0,0 +1,353 @@
+    virtual ~LDbTable();
+
+    bool Ok() const {return(m_IsOpen);};
+    leveldb::Iterator * NewIterator();
+
+    const leveldb::Status & GetStatus() const {return(m_LastStatus);};
+    const char * GetFileName() const {return(m_FileName.c_str());};
+
+    uint64_t GetSstCounter(unsigned Idx) const
+        {return(m_IsOpen ? m_TablePtr->GetSstCounters().Value(Idx) : 0);};
+
+protected:
+    leveldb::Options & m_Options;
+    std::string m_FileName;
+    leveldb::RandomAccessFile * m_FilePtr;
+    leveldb::Table * m_TablePtr;
+    uint64_t m_FileSize;
+    leveldb::Status m_LastStatus;
+
+    bool m_IsOpen;
+
+private:
+    // disable these
+    LDbTable();
+    LDbTable(const LDbTable &);
+    const LDbTable operator=(const LDbTable&);
+};  // LDbTable
+
+
+LDbTable::LDbTable(
+    leveldb::Options & Options,
+    std::string & FileName)
+    : m_Options(Options), m_FileName(FileName),
+      m_FilePtr(NULL), m_TablePtr(NULL), m_FileSize(0), m_IsOpen(false)
+{
+    m_LastStatus=m_Options.env->GetFileSize(m_FileName, &m_FileSize);
+
+    if (m_LastStatus.ok())
+        m_LastStatus=m_Options.env->NewRandomAccessFile(m_FileName, &m_FilePtr);
+
+    if (m_LastStatus.ok())
+    {
+        m_LastStatus=leveldb::Table::Open(m_Options, m_FilePtr, m_FileSize, &m_TablePtr);
+
+        // use fadvise to start file pre-read
+        m_FilePtr->SetForCompaction(m_FileSize);
+    }   // if
+
+    m_IsOpen=m_LastStatus.ok();
+
+    if (!m_IsOpen)
+    {
+        // some people would throw() at this point, but not me
+        delete m_TablePtr;
+        m_TablePtr=NULL;
+        delete m_FilePtr;
+        m_FilePtr=NULL;
+        m_FileSize=0;
+    }   // if
+
+    return;
+
+}   // LDbTable::LDbTable
+
+
+LDbTable::~LDbTable()
+{
+    m_IsOpen=false;
+    delete m_TablePtr;
+    m_TablePtr=NULL;
+    delete m_FilePtr;
+    m_FilePtr=NULL;
+    m_FileSize=0;
+}   // LDbTable::~LDbTable
+
+
+leveldb::Iterator *
+LDbTable::NewIterator()
+{
+    leveldb::Iterator * ret_ptr(NULL);
+
+    if (m_IsOpen)
+    {
+        leveldb::ReadOptions read_options;
+
+        read_options.fill_cache=false;
+        ret_ptr=m_TablePtr->NewIterator(read_options);
+    }   // if
+
+    return(ret_ptr);
+
+}   // LDbTable::NewIterator
+
+
+int
+main(
+    int argc,
+    char ** argv)
+{
+    bool error_seen, running, compare_files;
+    char ** cursor;
+
+    compare_files=false;
+    error_seen=false;
+    running=true;
+
+    // Options: needs filter & total_leveldb_mem initialized
+    leveldb::Options options;
+    options.filter_policy=leveldb::NewBloomFilterPolicy2(16);
+    options.total_leveldb_mem=(512 << 20);
+
+    // testing
+//    options.block_size=16 << 10;
+
+    for (cursor=argv+1; NULL!=*cursor && running; ++cursor)
+    {
+        // option flag?
+        if ('-'==**cursor)
+        {
+            char flag;
+
+            flag=*((*cursor)+1);
+            switch(flag)
+            {
+                //case 'b':  options.block_size=4096; break;
```

Comment:
Should probably add in some initial options to make this utility useful. Setting the block size and setting the compression method would both be good additions.